opengovsg/postmangovsg
# feat: Telegram webhook handler

## Problem

In order to send messages to Telegram recipients, we need to gather their phone number -> chat ID mapping. This is achieved by having users initiate a conversation with the bot and sending the bot their phone number. 

## Solution

This PR implements a serverless function that is able to:
- Handle incoming Telegram updates
- Lead the user through a short process to either subscribe to the bot or update their phone number
- Store the phone number -> chat ID mappings in the database for use by other Telegram-related components

**Features**:

The following Telegram commands are available to the user: 

- `/start` - Allows the user to subscribe to the bot
- `/updatenumber` - Allows the user to update their phone number -> chat ID mapping when they change their phone number

## Screenshots

`/start`

![Screen Shot 2020-06-08 at 12 49 12 PM](https://user-images.githubusercontent.com/6095637/83993665-a78d7980-a986-11ea-85a4-fb266569a120.png)

`/updatenumber`

![Screen Shot 2020-06-08 at 12 50 08 PM](https://user-images.githubusercontent.com/6095637/83993691-b7a55900-a986-11ea-8707-f3cce3546634.png)

## Tests

1. You will need a Telegram bot that you can control
    - Talk to `@BotFather` on Telegram to create one, note the bot token
2. Using `.env-example` as a reference, set `DEV_SERVER_BOT_TOKEN` in `.env` to your bot's token
3. Start the dev server with `npm run dev`
4. Use `ngrok` (or similar) to expose your dev server to the Internet (i.e. have a public-facing URL)
5. Set your bot's webhook URL to your dev server's public URL
    ```
    POST https://api.telegram.org/bot<bot token>/setWebhook

    {
      "url": "<public URL>/<bot id>"
    }
    ```
    `bot id` is the part before the `:` in your `bot token`
6. Start interacting with the bot – Test `/start` and `/updatenumber`
    - For `/updatenumber`, manually edit your phone number -> chat ID mapping in the db to something else, then go through the bot's update flow and verify that the mapping in the db is updated.

## Deploy Notes

- Ensure that Telegram-related tables exist and are up-to-date
- Deploy this as a lambda, setting the appropriate env vars as shown in `.travis.yml`
    The current setup only deploys on the `STAGING_TELEGRAM_HANDLER_BRANCH` branch. For prod, duplicate this block and replace with the relevant env vars. The Twilio staging+prod lambda deploy setup can be used as a reference.

**New dependencies**:

- `Telegraf`: Telegram Bot API library
